helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -144,19 +305,143 @@ public ZNRecord update(ZNRecord current) {
         return _record;
       }
-      // For users who are not the lock owner and try to do an update on a lock that is held by
-      // someone else, exception thrown is to be caught by data accessor, and return false for
-      // the update
-      LOG.error(
-          "User " + _userId + " tried to update the lock at " + new Date(System.currentTimeMillis())
-              + ". Lock path: " + _lockPath);
-      throw new HelixException("User is not authorized to perform this operation.");
+
+      // higher priority lock request will try to  preempt current lock owner
+      if (!isCurrentOwner(curLockInfo) && _priority > curLockInfo.getPriority()) {
+        if (curLockInfo.getRequestorId().equals("NONE")) {
+          // update lock Znode with requestor information
+          ZNRecord newRecord = composeNewRequestorRecord(curLockInfo, _record);
+          _lockStatus = LockConstants.LockStatus.PENDING;
+          _pendingTimeout =
+              _waitingTimeout > curLockInfo.getCleanupTimeout() ? curLockInfo.getCleanupTimeout()
+                  : _waitingTimeout;
+          return newRecord;
+        } // If the requestor field is not empty, and the coming lock request has a even higher
+        // priority. The new request will replace current requestor field of the lock
+        else if (_priority > curLockInfo.getRequestorPriority()) {
+          ZNRecord newRecord = composeNewRequestorRecord(curLockInfo, _record);
+          _lockStatus = LockConstants.LockStatus.PENDING;
+          long remainingCleanupTime =
+              curLockInfo.getCleanupTimeout() - (System.currentTimeMillis() - curLockInfo
+                  .getRequestingTimestamp());
+          _pendingTimeout = _waitingTimeout > remainingCleanupTime ? remainingCleanupTime :
+           _waitingTimeout;
+          return newRecord;
+        }
+      }
+
+      // For users who are not the lock owner and the priority is not higher than current lock
+      // owner, or the priority is higher than current lock, but lower than the requestor, throw
+      // an exception. The exception will be caught by data accessor, and return false for the
+      // update operation.
+      LOG.error("User {} failed to acquire lock at Lock path {}.", _userId, _lockPath);
+      throw new HelixException(
+          String.format("User %s failed to acquire lock at Lock path %s.", _userId, _lockPath));
     }
   }
 
+  /**
+   * Class that specifies how a lock node should be updated after the previous owner finishes the
+   * cleanup work
+   */
+  private class CleanupUpdater implements DataUpdater<ZNRecord> {
+    public CleanupUpdater() {
+    }
+
+    @Override
+    public ZNRecord update(ZNRecord current) {
+      // If we are still the lock owner, clean owner field.
+      LockInfo curLockInfo = new LockInfo(current);
+      if (isCurrentOwner(curLockInfo)) {
+        ZNRecord record = current;
```

Comment:
If I just create a new LockInfo(null) object, then I get the same result, right?
I guess maybe you want to make the LockInfo(<no parameter>) constructor public.

Reply:
Or even better, just define a static empty LockInfo object for this usage.

Reply:
Here we only clean up the owner related fields, but still contain requestor related fields, so that requestor can go ahead to acquire the lock. So we need to get the current lock and update based on that lock, instead of creating a default lock node.